embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt

```diff
@@ -1,14 +1,162 @@
+@file:Suppress("FunctionOnlyReturningConstant")
+
 package io.embrace.android.embracesdk.internal.delivery.scheduling
 
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.internal.comms.api.ApiResponse
+import io.embrace.android.embracesdk.internal.comms.api.Endpoint
 import io.embrace.android.embracesdk.internal.delivery.execution.RequestExecutionService
+import io.embrace.android.embracesdk.internal.storage.PayloadReference
+import io.embrace.android.embracesdk.internal.storage.StorageService2
+import io.embrace.android.embracesdk.internal.worker.BackgroundWorker
+import java.io.InputStream
+import java.util.Collections
+import java.util.LinkedList
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
 
 internal class SchedulingServiceImpl(
-    @Suppress("unused") private val requestExecutionService: RequestExecutionService
+    private val storageService: StorageService2,
+    private val executionService: RequestExecutionService,
+    private val schedulingWorker: BackgroundWorker,
+    private val deliveryWorker: BackgroundWorker,
+    private val clock: Clock
 ) : SchedulingService {
 
+    private val blockedEndpoints: MutableMap<Endpoint, Long> = ConcurrentHashMap()
+    private val sendLoopActive = AtomicBoolean(false)
+    private val activeSends: MutableSet<PayloadReference> = Collections.synchronizedSet(HashSet())
+    private val payloadsToRetry: MutableMap<PayloadReference, RetryInstance> = ConcurrentHashMap()
+
+    override fun onPayloadIntake() {
+        // When a payload arrives, check to see if there's already an active job try to deliver payloads
+        // If not, schedule job. If so, do nothing.
+        if (sendLoopActive.compareAndSet(false, true)) {
+            schedulingWorker.submit {
+                queueLoop()
+            }
+        }
+    }
+
     override fun handleCrash(crashId: String) {
+        // TODO: get ready to die
     }
 
-    override fun onPayloadIntake() {
+    /**
+     * Loop through the payloads ready to be sent by priority and queue for delivery
+     */
+    private fun queueLoop() {
+        try {
+            var readyPayloads = getReadyPayloads()
+            while (readyPayloads.isNotEmpty() && readyToSend()) {
+                readyPayloads.forEach { payload ->
+                    queueSend(payload)
+                }
+                readyPayloads = getReadyPayloads()
+            }
+        } finally {
+            sendLoopActive.set(false)
+            scheduleNextCheck()
+        }
+    }
+
+    private fun getReadyPayloads(): List<PayloadReference> {
+        val payloads = storageService.getPayloadsByPriority()
+        val readyPayloads = LinkedList<PayloadReference>()
+        payloads.forEach { payload ->
```

Comment:
So the initial thought was that `filter` will do the filtering at the beginning, but one condition for ineligibility is retry time - by the time it gets to a payload, it may be eligible for delivery. 

That said, if we keep the additional check at the end of the loop, any payloads that became eligible will be picked up then, so that will probably work just as well. Yeah, let me do that.